### DISCUSSION TITLE
SQLAlchemy + FastAPI Routers

### DISCUSSION THREAD

**Discussion by theinfosecguy** [fastapi/fastapi#6205] [Questions] [answered]
### Description

I am following the FastAPI integration guide with MySQL and it is working fine. The issue happens when I want to use the routers and want to divide the routes in multiple files. As of now, I have the route (which interacts with DB) in the main file but if I move this route to some other file say `routes.py` and import get_db function from the main file, it gives an error. 

Is there a way to use the DB function in all files just by importing?

``` 
# main.py

... imports....

def get_db:
    ......

@app.post("/users/", response_model=schemas.User)
def create_user(user: schemas.UserCreate, db: Session = Depends(get_db)):
    db_user = crud.get_user_by_email(db, email=user.email)
    if db_user:
        raise HTTPException(status_code=400, detail="Email already registered")
    return crud.create_user(db=db, user=user)
```

```
# routes.py
from .main import get_db

@app.post("/users/", response_model=schemas.User)
def create_user(user: schemas.UserCreate, db: Session = Depends(get_db)):
    db_user = crud.get_user_by_email(db, email=user.email)
    if db_user:
        raise HTTPException(status_code=400, detail="Email already registered")
    return crud.create_user(db=db, user=user)
```

Getting the following error:` ImportError: cannot import name 'get_db'`

### Operating System

macOS

### FastAPI Version

Latest

### Python Version

3.10

**Comment by mateoradman** [comment]
I suggest you create a separate file or a folder where you can store your dependencies such as `get_db` function that you mentioned above. This allows you to import it like:
```python
from .dependencies import get_db
```
Please take a look at @tiangolo's example https://github.com/tiangolo/full-stack-fastapi-postgresql/tree/master/%7B%7Bcookiecutter.project_slug%7D%7D/backend/app/app/api. More specifically, the file deps.py contains dependencies that are widely used for all endpoints.

**Comment by rafsaf** [comment, marked as answer]
It's more Python related than FastAPI related, you have a known and very common problem of circular imports in your code (most probably, because you wrote `... imports....` in main.py so I'm not 100% sure).

You can do as stated above (reasonable to move it to third file) or few other things.
There is great resource about this on RealPython:
https://realpython.com/python-import/#handle-cyclical-imports

You can read whole article as well.


As you are probably aware, there are thousands of open issues in FastAPI, if the comments answer your question, please close the issue, if they don't, don't hesitate to ask. Thanks 👍

**Comment by nhdinh** [comment]
There is a thing in mind: if the module is generally imported for other modules to work with, then it should be in a place where every other module can see it and can import it.
In your case, the get_db should be moved into a separate module, and then, other modules which contain routers can import it to use. For example
```
from .db import get_db
```

**Comment by TheJumpyWizard** [comment]
@theinfosecguy I would do as everyone else so far has suggestd to resolve dependency issues. I would move the get_db function into a separate file/directory. Separation of responsibilities is important especially as your code base grows, it will be easier to maintain. Typically I would just handle the responses in the API functions and delegate the db operations and data transformations to handlers that you call from the API

**Comment by theinfosecguy** [comment]
Thanks for the help! The issue is now resolved (by moving function to a new file).
